### PR TITLE
[shape_poly] Expand the support for shape polymorphism for jnp.pad

### PR DESCRIPTION
**Pull request by gnecula**
Handle several new padding modes: wrap, reflect, symmetric, linear_ramp, maximum. Not all situations are handled; try to give a clear error for the unsupported cases.

While implementing this, I needed to add shape polymorphism support also for jnp.linspace.

And I discovered a bug in the implementation of `divmod(0, b)`.

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -2685,13 +2685,14 @@ def _pad_wrap(array: Array, pad_width: PadValue[int]) -> Array:
       _check_no_padding(pad_width[i], "wrap")
       continue
     size = array.shape[i]
-    repeats, (left_remainder, right_remainder) = np.divmod(pad_width[i], size)
-    total_repeats = repeats.sum() + 1
+    left_repeats, left_remainder = divmod(pad_width[i][0], size)
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:2688]: I think we need to use `np.divmod` here in order for `left_remainder` to be static under JIT, otherwise the `if` statement below will lead to a concretization error. Maybe we're missing some test coverage?

Comment (gnecula) [reply jax/_src/numpy/lax_numpy.py:2688]: Here in normal JAX usage `pad_width` and `size` contain only constants (never tracers), and whether we use `np.divmod` or `divmod` the result will be constants. What I am doing here is enabling this code to work with symbolic dimensions, which `np.divmod` wants to cast to numerical values, while Python `divmod` is overloaded to handle symbolic dimensions.
